pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/functions/FunctionUtils.java

```diff
@@ -98,24 +95,29 @@ public static Functions searchForFunctions(String functionsDirectory, boolean al
         try (DirectoryStream<Path> stream = Files.newDirectoryStream(path, "*.nar")) {
             for (Path archive : stream) {
                 try {
-                    FunctionDefinition cntDef = FunctionUtils.getFunctionDefinition(archive.toString());
+
+                    NarClassLoader ncl = NarClassLoaderBuilder.builder()
+                            .narFile(new File(archive.toString()))
+                            .build();
+
+                    FunctionArchive.FunctionArchiveBuilder functionArchiveBuilder = FunctionArchive.builder();
+                    FunctionDefinition cntDef = FunctionUtils.getFunctionDefinition(ncl);
                     log.info("Found function {} from {}", cntDef, archive);
-                    log.error(cntDef.getName());
-                    log.error(cntDef.getFunctionClass());
+
+                    functionArchiveBuilder.archivePath(archive);
+
+                    functionArchiveBuilder.classLoader(ncl);
+                    functionArchiveBuilder.functionDefinition(cntDef);
+
                     if (alwaysPopulatePath || !StringUtils.isEmpty(cntDef.getFunctionClass())) {
-                        functions.functions.put(cntDef.getName(), archive);
+                        functions.put(cntDef.getName(), functionArchiveBuilder.build());
```

Comment:
When `alwayPopulatePath` == false or `cntDef.getFunctionClass()` is empty, function may be result empty. This is not the same as the original logic.  Please check if there is a problem, and add a unit test to cover it.

Reply:
The logic didn't change here (the condition is untouched). And there are already tests covering this. Or am I missing something ?

Reply:
https://github.com/apache/pulsar/pull/15693/files/2e9738aacb24e56571a0c7f9a629841b6290d532#r886692769